### PR TITLE
 createStream: Add functionality to create announcement streams. Part of #3131

### PR DESCRIPTION
**Pull request by jainkuniya**
Part of #3131

image

### CODE REVIEW

File: src/api/streams/createStream.js [hunk 1]
```diff
@@ -8,11 +8,13 @@ export default (
   description?: string = '',
```
Comment (borisyankov) [inline src/api/streams/createStream.js]: The last line of the commit message might be a bit premature :) (and can be omitted!)

File: src/streams/streamsActions.js [hunk 2]
```diff
@@ -43,6 +43,14 @@ export const updateExistingStream = (
   if (initialValues.invite_only !== newValues.isPrivate) {
     await updateStream(getAuth(getState()), id, 'is_private', newValues.isPrivate);
   }
+  if (initialValues.is_announcement_only !== newValues.isAnnouncementOnly) {
```
Comment (borisyankov) [inline src/streams/streamsActions.js]: This line of code looks OK but I am starting to be concerned we are calling `updateStream` too many times. Is there a 'batch update' version of this API call?

File: src/streams/CreateStreamScreen.js [hunk 3]
```diff
@@ -16,10 +16,15 @@ type Props = {
 class CreateStreamScreen extends PureComponent<Props> {
   props: Props;
 
-  handleComplete = (name: string, description: string, isPrivate: boolean) => {
+  handleComplete = (
+    name: string,
+    description: string,
+    isPrivate: boolean,
+    isAnnounceOnly: boolean,
```
Comment (borisyankov) [inline src/streams/CreateStreamScreen.js]: Should this be `isAnnouncementOnly`? (announce => announcement)

File: src/streams/EditStreamCard.js [hunk 4]
```diff
@@ -79,6 +91,11 @@ export default class EditStreamCard extends PureComponent<Props, State> {
           defaultValue={initialValues.invite_only}
           onValueChange={this.handleIsPrivateChange}
         />
+        <OptionRow
+          label="Restrict posting to organization administrators"
```
Comment (borisyankov) [inline src/streams/EditStreamCard.js]: We will need this in `messages_en.json` too.

**Comment by borisyankov** [comment]
Do we also need to expose the option of `history_public_to_subscribers`?

Note about English :) ... `it's === it is` and you mean `its` a few times in the commit messages.

**Comment by gnprice** [comment]
Thanks @jainkuniya ! I agree with all @borisyankov 's comments.

Another thought:
* This list of fields `name`, `description`, `inviteOnly`, and now `isAnnouncementOnly` appears in a lot of places, as your commit "createStream: Add functionality to ..." (btw "to to" -> "to") demonstrates. Also a lot of those places are positional arguments, and increasingly long lists of positional arguments make me nervous.
* How about we move these four into an object type of their own, like `StreamOptions`? Then all the places that pass around those four fields together can instead pass around one `streamOptions: StreamOptions`.
* That can come after this change. But it will be a good thing to do before we add a fifth option `historyPublicToSubscribers` :slightly_smiling_face: .

**Comment by borisyankov** [comment]
PS. The 'Deuglify' PR #3143 was inspired by these changes here and that the screen is getting busy.

@jainkuniya you should keep an eye on this PR :)

File: src/streams/CreateStreamScreen.js [hunk 5]
```diff
@@ -14,10 +14,17 @@ type Props = {|
 |};
 
 class CreateStreamScreen extends PureComponent<Props> {
-  handleComplete = (name: string, description: string, isPrivate: boolean) => {
+  props: Props;
```
Comment (borisyankov) [inline src/streams/CreateStreamScreen.js]: We no longer include `props: Props;`

File: src/streams/streamsActions.js (hunk 2 above)
Comment (jainkuniya) [reply src/streams/streamsActions.js]: submitted #3224 :)

File: src/streams/CreateStreamScreen.js (hunk 5 above)
Comment (jainkuniya) [reply src/streams/CreateStreamScreen.js]: Done.

I missed it while rebasing.

**Comment by gnprice** [comment]
Hi @jainkuniya ! I was just looking back through our P1 issues, and I see the one this PR is for. I'd love to pick it back up and get it merged. :slightly_smiling_face:

Would you rebase, and also take a look through and make any updates that seem like a good idea from looking at it again a while later? Then I'll give it a fresh review.

**Comment by gnprice** [comment]
There's some recent chat discussion in this thread:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Add.20functionality.20to.20create.20announcement.20streams/near/847205

**Comment by gnprice** [comment]
Thanks @jainkuniya ! Some comments from our discussion today:
* Let's call it something other than "picker", because it turns out that means something different from this in both the Apple and Google guidelines.

  There isn't a great standard name for this. But we could call it something like "SelectionList".

* Let's put them both in one file. They're so closely related -- each one only makes sense together with the other.

* Instead of Zulip green, let's make the text gray -- same as you see here: https://material.io/components/lists and in the Android system settings app.

* For the text of the options, use the same text that's in the webapp

* The spacing looks off. Like there's a bunch of space above "Private", and much less space between that and who-can-post, or between the latter and the "Create" button. Let's fix the spacing -- add more vertical margin or padding.

  Take a look at https://material.io/components/lists#specs and specifically "Two line", and check we're conforming to that.

* The alignment feels a bit off to me too, like the horizontal padding and margin should be adjusted as well. So it'd be good to check on that too.

---

A couple of further thoughts, on workflow:
* Let's make a fresh PR and close this one, because so much has changed from the version that was reviewed in this thread.
* In the next revision, please post screenshots.

**Comment by jainkuniya** [comment]
Github is not allowing me to create new PR without closing this, so closing this. And for reference will link other PR number in another comment

**Comment by jainkuniya** [comment]
This is moved to https://github.com/zulip/zulip-mobile/pull/4066

**Comment by gnprice** [comment]
> Github is not allowing me to create new PR without closing this

(I think specifically this is because you were using the same branch name.)

Closing this thread seems like the right thing, so :+1: -- discussion continues on #4066.

**Pull request closed without merging**